### PR TITLE
Feature/#20 suggestions

### LINKED ISSUE #20: Possibility to submit suggestions and other commands for moderators to mark them as "approved", "denied", etc
A command which lets users submit suggestions which will then be posted in a specific channels for people to vote on. Voting happens via two reactions. If enough up- or downvotes have been made and the difference between them is kinda significant, the embed of the suggestion should change its color. This signals if the idea is popular or not. 

Some other commands for moderators to mark suggestions as "approved", "denied", "considered" or "implemented".

### PR DESCRIPTION
**Pull request by PKlempe**
Closes #20

### CODE REVIEW

File: bot/feedback/feedback.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+        Args:
+            ctx (discord.ext.commands.Context): The context in which the command was called.
+            suggestion (str): The suggestion provided by the user.
+        """
+        await ctx.message.delete()
+        suggestion_id = self._db_connector.add_suggestion(ctx.author.id, ctx.message.created_at)
+
+        embed = _build_suggestion_embed(ctx.author, suggestion, suggestion_id)
+        message = await ctx.guild.get_channel(constants.CHANNEL_ID_SUGGESTIONS).send(embed=embed)
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:44]: get channel in init method

File: bot/feedback/feedback.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        corresponding embed will be adapted to represent this change and the user who submitted the idea will be
+        notified via DM.
+
+        Args:
+            ctx (discord.ext.commands.Context): The context in which the command was called.
+            suggestion_id (int): The id of the suggestion which should be approved.
+            reason (Optional[str]): The reason for this decision.
+        """
+        await ctx.message.delete()
+        await self._change_suggestion_status(SuggestionStatus.APPROVED, ctx.author, ctx.guild, suggestion_id, reason)
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:67]: I don't think the order of the arguments reflects their importance or the sense of the method. 

You could reorder it to (suggestion_id, suggestion_status, author, guild, reason). 
This would make it feel like you read: 
Change the suggestion Status of <id> to <suggestionstatusvalue> using/with help of author, guild and reason.

File: bot/feedback/feedback.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            ctx (discord.ext.commands.Context): The context in which the command was called.
+            suggestion_id (int): The id of the suggestion which should be approved.
+            reason (Optional[str]): The reason for the decision.
+        """
+        await ctx.message.delete()
+        await self._change_suggestion_status(SuggestionStatus.CONSIDERED, ctx.author, ctx.guild, suggestion_id, reason)
+
+        log.info("Suggestion #%s is being considered by %s.", suggestion_id, ctx.author)
+
+    @manage_suggestions.command(name="implemented")
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:111]: I am not sure if it is better to have 5 different methods for the status or to just have one and parse the status as an argument.
Maybe if you specify the type to SuggestionStatus in the parameter list and make sure it has the right values, it might even auto-parse.

File: bot/feedback/feedback.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        """
+        if isinstance(error, commands.BadArgument):
+            await ctx.send("Ich konnte leider keinen Vorschlag mit der von dir angegebenen Nummer finden. :frowning2:",
+                           delete_after=constants.TIMEOUT_INFORMATION)
+
+    async def _change_suggestion_status(self, status: SuggestionStatus, author: discord.Member, guild: discord.Guild,
+                                        suggestion_id: int, reason: Optional[str]):
+        """Method which changes the status of a suggestion.
+
+        Changes the status of the suggestion in our db, adapts the corresponding embed in the suggestion channel to
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:153]: in *the* db

File: bot/feedback/feedback.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        """
+        self.bot = bot
+        self._db_connector = DatabaseConnector(constants.DB_FILE_PATH, constants.DB_INIT_SCRIPT)
+
+    @commands.group(name="suggestion", hidden=True, invoke_without_command=True, aliases=["suggest"])
+    @command_log
+    async def manage_suggestions(self, ctx: commands.Context, *, suggestion: str):
+        """Command Handler for the `suggestion` command.
+
+        Allows users to submit suggestions for improving our server. After submitting, an embed containing the provided
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:31]: *the* server

File: bot/feedback/feedback.py (hunk 4 above)
Comment (SebastiantheAstorian) [reply bot/feedback/feedback.py:153]: Consider also other occurences of 'our'.

File: bot/feedback/feedback.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        Args:
+            status (SuggestionStatus): The new status of the suggestion.
+            author (discord.Member): The user who invoked the command to change it.
+            guild (discord.Guild): The Discord server on which to find the suggestion.
+            suggestion_id (int): The id of the suggestion.
+            reason (Optional[str]): The reason for the decision.
+        """
+        id_exists = self._db_connector.set_suggestion_status(suggestion_id, status)
+        if not id_exists:
+            raise commands.BadArgument
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:165]: Pass a message with the Error. 
Also I believe you do not raise an Instance of BadArgument, but an instance of type with the value <class: BadArgument> (i think you forgot the braces).

File: bot/feedback/feedback.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+        suggestion_data = self._db_connector.get_suggestion(suggestion_id)
+        message = await guild.get_channel(constants.CHANNEL_ID_SUGGESTIONS).fetch_message(suggestion_data[0])
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:168]: What happens if the msg has been deleted? Did we ever consider this before? Is this a problem?

File: bot/feedback/feedback.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        emojis specified in constants.py, the color of the suggestion embed will be changed if enough up- or downvotes
+        have been submitted.
+
+        Args:
+            payload (discord.RawReactionActionEvent): The payload for the triggered event.
+        """
+        if not payload.member.bot and payload.channel_id == constants.CHANNEL_ID_SUGGESTIONS and \
+                self._db_connector.get_suggestion_status(payload.message_id) == SuggestionStatus.UNDECIDED:
+            message = await self.bot.get_channel(payload.channel_id).fetch_message(payload.message_id)
+            reactions = message.reactions
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:187]: Is it technically possible to write:

```
( await self.bot.get_channel(payload.channel_id).fetch_message(payload.message_id) ).reactions
```

File: bot/feedback/feedback.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+            if payload.emoji.name in (constants.EMOJI_UPVOTE, constants.EMOJI_DOWNVOTE):
+                total_votes = reactions[0].count + reactions[1].count
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:190]: You could directly use
`required_difference= (reactions[0].count + reactions[1].count) /2`

File: bot/feedback/feedback.py [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+                # Changes the color of the embed depending if one side received at least 10 votes and the difference
+                # between them is bigger than 50% of total votes.
+                if constants.LIMIT_SUGGESTION_VOTES < reactions[0].count > reactions[1].count and \
+                        difference > (total_votes / 2):
+                    new_embed = _recolor_embed(message.embeds[0], constants.EMBED_COLOR_SUGGESTION_MEMBERS_LIKE)
+                    await message.edit(embed=new_embed)
+                elif constants.LIMIT_SUGGESTION_VOTES < reactions[1].count and difference > (total_votes / 2):
+                    new_embed = _recolor_embed(message.embeds[0], constants.EMBED_COLOR_SUGGESTION_MEMBERS_DISLIKE)
+                    await message.edit(embed=new_embed)
+
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:202]: You can pull the `await message.edit(embed=new_embed)` out of the if/else block

File: bot/feedback/feedback.py [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        author (discord.Member): The user who invoked the command to change it.
+        reason (Optional[str]): The reason for the decision.
+        status (SuggestionStatus): The new status of the suggestion.
+    """
+    dict_embed = message.embeds[0].to_dict()
+
+    if "fields" in dict_embed:
+        dict_embed["fields"].clear()
+    else:
+        dict_embed["fields"] = list()
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:239]: Why do you need to make a distinction and not just always replace it with an empty list?

File: bot/feedback/feedback.py [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+    if reason:
+        dict_embed["fields"].append({"name": f"Begründung von {author}:", "value": reason, 'inline': False})
+    else:
+        dict_embed = message.embeds[0].to_dict()
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:244]: Replace all of this with:

```
dict_embed["fields"] = [{"name": f"Begründung von {author}:", "value": reason, 'inline': False}] if reason else list()
```

File: bot/feedback/feedback.py [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+    dict_embed["title"] = dict_embed["title"].split(" -")[0]
+    if status == SuggestionStatus.APPROVED:
+        dict_embed["color"] = constants.EMBED_COLOR_SUGGESTION_APPROVED
+        dict_embed["title"] = dict_embed["title"] + " - Genehmigt :white_check_mark:"
+    elif status == SuggestionStatus.DENIED:
+        dict_embed["color"] = constants.EMBED_COLOR_SUGGESTION_DENIED
+        dict_embed["title"] = dict_embed["title"] + " - Abgelehnt :no_entry_sign:"
+    elif status == SuggestionStatus.CONSIDERED:
```
Comment (SebastiantheAstorian) [inline bot/feedback/feedback.py:253]: Could you possibly assign the color and Text to the enum (for example as a tuple) and just write:

```
# suggestion_status.py
class SuggestionStatus(enum):
    ...
    DENIED = (2, constants.EMBED_COLOR_SUGGESTION_DENIED, " - Abgelehnt :no_entry_sign:")
    ....

#feedback.py (the commented code)
dict_embed["color"] = status[1]
dict_embed["title"] = dict_embed["title"] + status[2]
```
No need for if/else then

File: bot/persistence/database_connector.py [hunk 14]
```diff
@@ -57,6 +58,100 @@ def remove_module_role(self, role_id: int):
...
+
+        Args:
+            author_id (int): The id of the user who submitted the suggestion.
+            timestamp (datetime.datetime): A timestamp when this suggestion has been submitted.
+
+        Returns:
+            int: The row id of the new entry.
+        """
+        with DatabaseManager(self._db_file) as db_manager:
+            row_id = db_manager.execute(queries.INSERT_SUGGESTION, (author_id, timestamp)).lastrowid
```
Comment (SebastiantheAstorian) [inline bot/persistence/database_connector.py:72]: Is lastrowid the row number or actually the value in the id column. 

Because any vacuuming or similar things in the db will change the row numbers.

File: bot/persistence/database_connector.py [hunk 15]
```diff
@@ -57,6 +58,100 @@ def remove_module_role(self, role_id: int):
...
+
+    def set_suggestion_status(self, suggestion_id: int, status: SuggestionStatus) -> bool:
+        """Sets the status of a suggestion with the specified id.
+
+        Args:
+            suggestion_id (int): The id of the suggestion.
+            status (SuggestionStatus): The new status of the suggestion.
+
+        Returns:
+            bool: A boolean representing if a row has been changed.
```
Comment (SebastiantheAstorian) [inline bot/persistence/database_connector.py:130]: ... *if any rows have been changed*

File: bot/persistence/resources/init_db.sql [hunk 16]
```diff
@@ -5,6 +5,14 @@ CREATE TABLE IF NOT EXISTS Modmail(
     Timestamp TEXT NOT NULL
 );
 
+CREATE TABLE IF NOT EXISTS Suggestion(
+    ID INTEGER PRIMARY KEY,
+    MessageID TEXT,
+    AuthorID TEXT NOT NULL,
+    StatusID SMALLINT NOT NULL DEFAULT 0,
+    Timestamp TEXT NOT NULL
```
Comment (SebastiantheAstorian) [inline bot/persistence/resources/init_db.sql:13]: If we don't need this, remove it. Also we have the timestamp on the message.

File: bot/university/university.py [hunk 17]
```diff
@@ -251,10 +252,7 @@ def check_reaction(_reaction, user):
 
     @commands.group(name='exchange', hidden=True, invoke_without_command=True)
     @command_log
-    async def exchange(self,
-                       ctx: commands.Context,
-                       channel: discord.TextChannel,
-                       offered_group: int,
+    async def exchange(self, ctx: commands.Context, channel: discord.TextChannel, offered_group: int,
```
Comment (SebastiantheAstorian) [inline bot/university/university.py:255]: :(

**Review by SebastiantheAstorian — CHANGES_REQUESTED** [review]

File: bot/feedback/feedback.py (hunk 3 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:111]: Doesn't work because the suggestion command has a keyword-only argument which would make it really hard to parse all of the arguments. I could make another subcommand like `set` or something like that, but then the command would only get longer. Example: `!suggestion set approve` instead of simply `!suggestion approve`

File: bot/feedback/feedback.py (hunk 7 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:168]: It depends. `fetch_message()` should return `None` which then results in an exception like `NoneType doesn't have method xxxx`. This will be logged but the bot shouldn't be interrupted or even crash because of something like that.

File: bot/feedback/feedback.py (hunk 8 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:187]: Doesn't work because we have to await the channel.

File: bot/feedback/feedback.py (hunk 10 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:202]: Yeah but then Discord would refresh the embed everytime someone adds a reaction to it. So the point of this line being inside the if/elif is that the message will only be changed (and therfore reloaded) if the specified criteria is met.

File: bot/feedback/feedback.py (hunk 11 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:239]: Good point. Idk, maybe I was tired.

File: bot/feedback/feedback.py (hunk 12 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:244]: Would also remove the `inline: False` because it doesn't matter if the list gets cleared every time.

File: bot/feedback/feedback.py (hunk 13 above)
Comment (PKlempe) [reply bot/feedback/feedback.py:253]: I could do that but then it would break the code where I get a value from the DB and try to get an enum constant with it. For this to work I would either have to use the complete tuple to get the constant or write some method which returns the constant by providing the first value of the tuple. Either way I don't think this is needed just because some if/elif.

File: bot/persistence/database_connector.py (hunk 14 above)
Comment (PKlempe) [reply bot/persistence/database_connector.py:72]: > If a table contains a column of type INTEGER PRIMARY KEY, then that column becomes an alias for the ROWID. You can then access the ROWID using any of four different names, the original three names described above or the name given to the INTEGER PRIMARY KEY column. **All these names are aliases for one another and work equally well in any context.**

https://sqlite.org/autoinc.html

File: bot/persistence/resources/init_db.sql (hunk 16 above)
Comment (PKlempe) [reply bot/persistence/resources/init_db.sql:13]: It would be possible that I need this timestamp in the future. I'm considering if I should implement a commands which returns me a list of all suggestions which have no decision or which are currently being considered. When I do that, I would like to have the timestampt to see since when the suggestion is "open" or something. Of course I could use the timestamp from the message itself but I tried that with the modmail before and the API calls really slowed things down. That's why I store the timestamp for the modmail in the db too.

**Review by SebastiantheAstorian — APPROVED** [review]
